rero_ils/modules/patrons/api.py

```diff
@@ -775,6 +779,52 @@ def get_current_patron(record):
             if ptrn.organisation_pid == record.organisation_pid:
                 return ptrn
 
+    @classmethod
+    def set_communication_channel(
+            cls, user=None, dbcommit=True, reindex=True):
+        """Set communication_channel for patrons according to patrons data.
+
+        :param user - user record
+        :param dbcommit - commit the changes
+        :param reindex - index the changes
+        """
+        def basic_query(channel):
+            """Returns basic ES query."""
+            return PatronsSearch()\
+                .filter('term', user_id=user.id)\
+                .filter('term', patron__communication_channel=channel)\
+                .source(includes='pid')\
+
+        mail_query = basic_query(CommunicationChannel.EMAIL)\
+            .filter('bool', must_not=[
+                Q('exists', field='patron.additional_communication_email')])\
+            .filter('bool', must_not=[Q('exists', field='email')])
+        to_mail_pids = [[hit['pid'], CommunicationChannel.MAIL, hit.meta.id]
+                        for hit in mail_query.scan()]
+        email_query = basic_query(CommunicationChannel.MAIL)\
+            .filter('bool', should=[
+                Q('exists', field='patron.additional_communication_email'),
+                Q('exists', field='email')
+            ])
+        to_email_pids = [[hit['pid'], CommunicationChannel.EMAIL, hit.meta.id]
+                         for hit in email_query.scan()]
+        pids = to_mail_pids + to_email_pids
+        ids = []
+        for pid, channel, id in pids:
+            ids.append(id)
+            patron = Patron.get_record_by_pid(pid)
+            if patron:
+                patron['patron']['communication_channel'] = channel
```

Comment:
I would prefer the `update` method of the api with validation.

Reply:
As discussed, no need to add delay to the user editor because of this.